src/Bicep.Core/Emit/TemplateWriter.cs

```diff
@@ -77,53 +79,33 @@ private static string GetSchema(ResourceScope targetScope)
 
             return "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#";
         }
-
-        private readonly JsonTextWriter writer;
         private readonly EmitterContext context;
-
         private readonly string assemblyFileVersion;
 
-        public TemplateWriter(JsonTextWriter writer, SemanticModel semanticModel, string assemblyFileVersion)
+        public TemplateWriter(SemanticModel semanticModel, string assemblyFileVersion)
         {
-            this.writer = writer;
             this.context = new EmitterContext(semanticModel);
             this.assemblyFileVersion = assemblyFileVersion;
         }
 
-        /// <summary>
-        /// This emits the template into memory, calculates the template hash of the template
-        /// It then adds the templateHash field to the template and finally writes it to our desired JsonTextWriter
-        /// </summary>
-        public void Write()
+        public void Write(JsonTextWriter writer)
         {
-            string templateString;
-            using (StringWriter stringWriter = new())
-            using (JsonTextWriter memoryWriter = new(stringWriter))
-            {
-                var emitter = new ExpressionEmitter(memoryWriter, this.context);
-                // no templatehash, write to memory
-                this.Write(memoryWriter, emitter, emitMetadata: true);
-                // TODO: avoid reading the whole template into a string. Address this when ComputeTemplateHash 
-                // has a streaming variant.
-                templateString = stringWriter.ToString();
-            }
-            var templateHash = TemplateHashExtensions.ComputeTemplateHash(templateString);
-            JObject template = (JObject)JObject.Parse(templateString);
-            ((JObject) template["metadata"]!["_generator"]!).Add(new JProperty("templateHash", templateHash));
-            template.WriteTo(this.writer);
+            JToken template;
+            using (StringWriter stringWriter = new ())
+            {
+                this.Write(stringWriter, writer);
+                template = stringWriter.ToString().FromJson<JToken>();
+            }
+            var templateHash = TemplateHelpers.ComputeTemplateHash(template);
+            ((JObject) template.SelectToken(GeneratorMetadataPath)!).Add(new JProperty("templateHash", templateHash));
+            template.WriteTo(writer);
         }
 
-        /// <summary>
-        /// Writing modules behaves differently than Write: we reuse the parent memoryWriter and do not generate metadata
-        /// </summary>
-        private void WriteModule()
+        private void Write(StringWriter source, JsonTextWriter destination)
```

Comment:
Does destination get used here at all?

I'd suggest a contract of either:
```csharp
private void WriteWithoutHash(JsonTextWriter writer)
```
Or (since you're converting it to a JToken anyway):
```csharp
private JToken GenerateTemplateWithoutHash()
```